src/client/datascience/variablesView/variableViewProvider.ts

```diff
@@ -17,6 +19,24 @@ import { VariableView } from './variableView';
 export class VariableViewProvider implements IVariableViewProvider {
     public readonly viewType = 'jupyterViewVariables';
 
+    // Either return the active variable view or wait until it's created and return it
+    public get activeVariableView(): Promise<VariableView> {
+        // For test execution only
```

Comment:
This is one of the main areas that I'm still mulling over some. Both this provider and the webview host now exposed some functions that should only be used by test code. I'm a bit leery of this, but had the following reasoning:

1. Wrapping / patching options would be adding more complexity
2. Keeping the code that we test as much 'exact product code' as possible felt good for the .vscode test
3. Only a few functions were needed, and they could be marked with isTestExecution to prevent product usage

On the downside, it's exposing a few things that don't need to be exposed in non-test scenarios. Opinions on this? And on the HTML fetching code in the webview? If there are ideas for a cleaner way to hide this without adding a bunch of test -> product code complexity I'd be down for it.

Reply:
It's just typescript check so I guess I could keep them private and then cast to a test specific interface in the test code, not sure that's any better.

Reply:
Making the isTestExecution bit a decorator might also work.

Reply:
I've done the equivalent of the following (involves the least typing, in both senses of the word). Maybe bad practice but nobody mentioned it on the code reviews:
```typescript
   (variableViewProvider as any).activeVariableView
```

But the more consistent thing to do would probably be the test interface idea. We do this for IExtensionTestApi as well:

https://github.com/microsoft/vscode-jupyter/blob/e8f311aa27269c2433d00a04f90cfd18c803f463/src/test/initialize.ts#L31

Reply:
I like the extension test api that exposes privates only in tests.

So I believe here this would be IVariableViewTestProvider or something like that and it would have:

```
interface IVariableViewTestProvider {
   readonly activeVariableView: Promise<IVariableView>
}
```

Reply:
@joyceerhl Yeah, leaving them dangling out in public doesn't seem good. Maybe the following?
1. Decorator for the isTestExecution check (I like this as it also annotates the function)
2. Interface for access to the test specific functions, Interfaces live in the test code. Keeping them in the test code prevents them from being seen as "oh, maybe I can use this" when looking at product code.
3. Us the (X as any) as Y cast in test code to get access to the private test methods.

Reply:
The real class wouldn't actually implement this but would just leave the activeVariableView as a private Although that makes it a tad breakable.

You could also do like we did for the InteractiveWindowProvider where in tests, a different provider is added to the container. The test provider implements the IVariableViewTestProvider so then at least there's compile protection for the activeVariableView.